File: ruby/060_Chimera.rb
# 060 - Chimera（★5）
# 最長増加部分列(LIS), 二分探索

N = gets.to_i
A = gets.split.map(&:to_i)

# 返す配列のi番目はi番目の値を最大とするLISの長さ
# dpは短調増加だが、返す配列は短調増加ではない。
# arr自体の最長増加部分列の長さはdpのsize or 返り値の最大値で求まる
def lis(arr)
  n = arr.size
  dp = []
  arr.reduce([]) do |result, a|
    pos = dp.bsearch_index { _1 >= a } || dp.size
    dp[pos] = a
    result << pos + 1
  end
end

P = lis(A)
Q = lis(A.reverse)
puts N.times.map { P[_1] + Q[N - _1 - 1] - 1 }.max

File: kotlin/060_Chimera.kt
/**
 * 060 - Chimera（★5）
 * 最長増加部分列(LIS)
 * 二分探索
 */

fun main() {
  val N = readLine()!!.toInt()
  val A = readLine()!!.split(" ").map(String::toInt)
  val P = lis(A)
  val Q = lis(A.reversed())

  val max = (0 until N).map { P[it] + Q[N - it - 1] - 1 }.max()
  println(max)
}

/**
 * 返す配列のi番目はi番目の値を最大とするLISの長さ
 * dpは短調増加だが、返す配列は短調増加ではない。
 * arr自体の最長増加部分列の長さはlisMemoのsize or 返り値の最大値で求まる
 */
fun lis(arr: List<Int>): MutableList<Int> {
  val lisMemo = mutableListOf<Int>()
  val result = mutableListOf<Int>()

  for (i in arr) {
    val idx = lisMemo.binarySearch(i).let { if (it >= 0) it else it.inv() }
    if (idx == lisMemo.size) lisMemo.add(i) else lisMemo.set(idx, i)
    result.add(idx + 1)
  }
  return result
}
